spring-context/src/main/java/org/springframework/context/ApplicationEvent.java

```diff
@@ -46,6 +47,17 @@ public ApplicationEvent(Object source) {
 		this.timestamp = System.currentTimeMillis();
 	}
 
+	/**
+	 * Create a new {@code ApplicationEvent} with a fixed timestamp.
+	 * @param source the object on which the event initially occurred or with
+	 * which the event is associated (never {@code null})
+	 * @param clock a clock which will provide the timestamp
+	 */
+	public ApplicationEvent(Object source, Clock clock) {
+		super(source);
+                this.timestamp = clock.millis();
```

Comment:
```suggestion
		this.timestamp = clock.millis();
```

Tabs instead of spaces, please.